### PR TITLE
SALTO-2054: extract tags to references

### PR DESCRIPTION
**Pull request by noamhammer**
_extract tags to references_

---

_Additional context for reviewer_
* Tags are not objects in zendesk (they don't have internal id and such)
* There is an API to list the 500 most popular tags (which doesn't help us)
* We decided to extract it based on all the places that refer to tags
* In those places, tags are defined with space as a separator
* There is no deploy for tags - so I created an empty deploy function

---
_Release Notes_: 
_Zendesk_support adapter:_
* Add references to tags

---
_User Notifications_: 
_New tags instances will be added and the places that refer to them will be changed_

### CODE REVIEW

File: packages/zendesk-support-adapter/src/filters/tag.ts [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+  value.split(TAG_SEPERATOR).filter(tag => !_.isEmpty(tag))
+
+const replaceTagsWithReferences = (instance: InstanceElement): string[] => {
+  const tags: string[] = []
+  RELEVANT_CONDITION_FIELD_NAMES
+    .forEach(fieldName => {
+      const conditions = _.get(instance.value, fieldName)
+      const fullName = instance.elemID
+        .createNestedID(...fieldName.split(ElemID.NAMESPACE_SEPARATOR)).getFullName()
+      if (!areConditions(conditions, fullName)) {
```
Comment (netama) [inline packages/zendesk-support-adapter/src/filters/tag.ts:54]: (continuing from the comment on the function implementation) is this a valid/expected scenario for some of the types? if so, we shouldn't log a warning when we reach an irrelevant type (e.g. it seems we will log a warning whenever there is no value at all, which will pollute the logs?)

Comment (netama) [inline packages/zendesk-support-adapter/src/filters/tag.ts:54]: (continuing from the comment on the function implementation) is this a valid/expected scenario for some of the types? if so, we shouldn't log a warning when we reach an irrelevant type (e.g. it seems we will log a warning whenever there is no value at all, which will pollute the logs?)

File: packages/zendesk-support-adapter/src/filters/tag.ts [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+const extractTags = (value: string): string[] =>
+  value.split(TAG_SEPERATOR).filter(tag => !_.isEmpty(tag))
+
+const replaceTagsWithReferences = (instance: InstanceElement): string[] => {
+  const tags: string[] = []
+  RELEVANT_CONDITION_FIELD_NAMES
+    .forEach(fieldName => {
+      const conditions = _.get(instance.value, fieldName)
+      const fullName = instance.elemID
+        .createNestedID(...fieldName.split(ElemID.NAMESPACE_SEPARATOR)).getFullName()
```
Comment (netama) [inline packages/zendesk-support-adapter/src/filters/tag.ts:53]: can we split the lines in a more consistent way? (break on any `.` or at least move `.getFullName()` below)?

Comment (netama) [inline packages/zendesk-support-adapter/src/filters/tag.ts:53]: can we split the lines in a more consistent way? (break on any `.` or at least move `.getFullName()` below)?

File: packages/zendesk-support-adapter/src/filters/tag.ts [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+const RELEVANT_TYPE_NAMES = [
+  'automation',
+  'trigger',
+  'view',
+  'macro',
+  'sla_policy',
+  'workspace',
+]
+const RELEVANT_FIELD_NAMES = ['current_tags', 'remove_tags', 'set_tags']
+const RELEVANT_CONDITION_FIELD_NAMES = ['actions', 'conditions.all', 'conditions.any', 'filter.all', 'filter.any']
```
Comment (netama) [inline packages/zendesk-support-adapter/src/filters/tag.ts:37]: can we have these field names mapped by type? or are they all the same in all types?

Comment (netama) [inline packages/zendesk-support-adapter/src/filters/tag.ts:37]: can we have these field names mapped by type? or are they all the same in all types?

File: packages/zendesk-support-adapter/src/filters/tag.ts [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+      const fullName = instance.elemID
+        .createNestedID(...fieldName.split(ElemID.NAMESPACE_SEPARATOR)).getFullName()
+      if (!areConditions(conditions, fullName)) {
+        return
+      }
+      conditions.forEach(condition => {
+        if (
+          RELEVANT_FIELD_NAMES.includes(condition.field)
+          && _.isArray(condition.value)
+          && condition.value.every(_.isString)
```
Comment (netama) [inline packages/zendesk-support-adapter/src/filters/tag.ts:84]: are all failure modes valid? or do we need two `if` levels with a warning?

Comment (netama) [inline packages/zendesk-support-adapter/src/filters/tag.ts:84]: are all failure modes valid? or do we need two `if` levels with a warning?

File: packages/zendesk-support-adapter/src/filters/tag.ts [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+        ) {
+          condition.value = condition.value.join(TAG_SEPERATOR)
+        }
+      })
+    })
+}
+
+/**
+ * Add dependencies from elements to dynamic content items in
```
Comment (netama) [inline packages/zendesk-support-adapter/src/filters/tag.ts:93]: please update for this filter :)

Comment (netama) [inline packages/zendesk-support-adapter/src/filters/tag.ts:93]: please update for this filter :)

File: packages/zendesk-support-adapter/src/filters/tag.ts [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+ * the _generated_ dependencies annotation
+ */
+const filterCreator: FilterCreator = ({ config }) => ({
+  onFetch: async (elements: Element[]): Promise<void> => {
+    const instances = elements
+      .filter(isInstanceElement)
+      .filter(instance => RELEVANT_TYPE_NAMES.includes(instance.elemID.typeName))
+
+    const tags = instances.map(instance => replaceTagsWithReferences(instance)).flat()
```
Comment (netama) [inline packages/zendesk-support-adapter/src/filters/tag.ts:102]: 1. there are side-effects here, can we avoid them by creating new instances?
2. do we need to change the fields' type definitions (from string to string array)?

Comment (netama) [inline packages/zendesk-support-adapter/src/filters/tag.ts:102]: 1. there are side-effects here, can we avoid them by creating new instances?
2. do we need to change the fields' type definitions (from string to string array)?

File: packages/zendesk-support-adapter/src/filters/tag.ts [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+          naclCase(tag),
+          tagObjectType,
+          { id: tag },
+          [ZENDESK_SUPPORT, RECORDS_PATH, TAG_TYPE_NAME, pathNaclCase(tag)]
+        ))
+      .value();
+    // We do this trick to avoid calling push with the spread notation
+    [tagObjectType, tagInstances].flat().forEach(element => { elements.push(element) })
+  },
+  deploy: async (changes: Change<InstanceElement>[]) => {
```
Comment (netama) [inline packages/zendesk-support-adapter/src/filters/tag.ts:123]: I think we should document very clearly that we are not deploying anything

Comment (netama) [inline packages/zendesk-support-adapter/src/filters/tag.ts:123]: I think we should document very clearly that we are not deploying anything

**Comment by netama** [comment]
@noamhammer we need a noise-reduction PR for this

File: packages/adapter-components/src/deployment/change_validators/check_deployment_based_on_config.ts [hunk 8]
```diff
@@ -60,6 +62,9 @@ export const createCheckDeploymentBasedOnConfigValidator = (
         const typeConfig = apiConfig?.types?.[typeName]?.deployRequests ?? {}
         return createChangeErrors(typeConfig, element.elemID, change.action)
       }
+      if (typesWithNoDeploy.includes(element.elemID.typeName)) {
+        return []
```
[the remark this answers is not in the document]
Comment (noamhammer) [reply packages/adapter-components/src/deployment/change_validators/check_deployment_based_on_config.ts:66]: Why? this is the happy flow for some types

[the remark this answers is not in the document]
Comment (noamhammer) [reply packages/adapter-components/src/deployment/change_validators/check_deployment_based_on_config.ts:66]: Why? this is the happy flow for some types

File: packages/zendesk-support-adapter/src/filters/tag.ts [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+
+const { RECORDS_PATH, TYPES_PATH } = elementsUtils
+const { awu } = collections.asynciterable
+
+const TAG_SEPERATOR = ' '
+const extractTags = (value: string): string[] =>
+  value.split(TAG_SEPERATOR).filter(tag => !_.isEmpty(tag))
+
+const replaceTagsWithReferences = (instance: InstanceElement): string[] => {
+  const tags: string[] = []
```
[the remark this answers is not in the document]
Comment (noamhammer) [reply packages/zendesk-support-adapter/src/filters/tag.ts:48]: Talked f2f, I don't agree with you on that one

File: packages/zendesk-support-adapter/src/filters/utils.ts [hunk 10]
```diff
@@ -65,3 +71,17 @@ export const createAdditionalParentChanges = async (
...
+
+const EXPECTED_CONDITION_SCHEMA = Joi.array().items(Joi.object({
+  field: Joi.string().required(),
+  value: Joi.optional(),
+}).unknown(true)).required()
+
+export const areConditions = (values: unknown, fullName: string): values is Condition[] => {
+  const { error } = EXPECTED_CONDITION_SCHEMA.validate(values)
+  if (error !== undefined) {
+    log.warn(`Received an invalid values for conditions on ${fullName}: ${safeJsonStringify(values)}`)
```
[the remark this answers is not in the document]
Comment (noamhammer) [reply packages/zendesk-support-adapter/src/filters/utils.ts:83]: re: 1 - you are correct
re: 2 - the are/is<TypeName> is our current convention - both in Zendesk and in Jira - so I will keep it like that

**Review by netama — APPROVED** [review]
LGTM once the noise-reduction PR is merged

**Pull request merged**